docs/howtos/DataPlane Generation - DPG/07tcgcTypes.mdx

```diff
@@ -0,0 +1,1477 @@
+This allows for the circular nature of an `SdkModelPropertyType`.
+
+```
+import { SdkServiceOperation, SdkModelPropertyType } from "@azure-tools/typespec-client-generator-core";
+import { PythonSdkContext } from "./lib.js";
+import { getPythonSdkType } from "./types.js";
+
+export function getSdkModelPropertyType<
+  TServiceOperation extends SdkServiceOperation
+>(
+  context: PythonSdkContext<TServiceOperation>,
+  prop: SdkModelPropertyType
+): PythonSdkModelType {
+  const type = getPythonSdkType(context, prop.type)
+  switch (prop.kind) {
+    case "query":
+    case "header"
+    case "path":
+      return {
+        ...prop,
+        type,
+        correspondingMethodParams: prop.correspondingMethodParams.map(
+          (x) => getSdkModelPropertyType(context, x),
+        )
+      }
+    default:
+      return {
+        ...prop,
+        type,
+      }
+  }
+}
+```
+
+### SdkMethodResponse
+
+This represents the response that our method will ultimately return.
+It is not tied to a transport, like http, it solely represents what we want to return to users
+
+#### Interface
+
+```ts
+export interface SdkMethodResponse {
+  kind: "method";
+  // the path from the service response to what the method returns.
+  logicalPath?: string;
+  type?: SdkType;
+  description?: string;
+  details?: string;
+}
+```
+
+#### Example
+
+```ts
+const sdkMethodResponse = {
+  kind: "method",
+  logicalPath: undefined,
+  type: {
+    kind: string,
+  },
+}
+
+const sdkPagingMethodResponse = {
+  kind: "method",
+  logicalPath: "value",
+  type: {
+    kind: "array",
+    valueType: {
+      kind: "model",
+      name: "Widget",
+      ...
+    }
+  }
+}
+```
+
+#### Usage
+
+```ts
+import { SdkMethodResponse } from "@azure-tools/typespec-client-generator-core";
+import { PythonSdkContext, PythonSdkMethodResponse } from "./lib.js";
+import { getPythonSdkType } from "./types.js";
+
+function serializeMethodResponse<
+  TServiceOperation extends SdkServiceOperation
+>(
+  context: PythonSdkContext<TServiceOperation>,
+  response: SdkMethodResponse,
+): PythonSdkMethodResponse {
+  return {
+    ...response,
+    type: getPythonSdkType(context, response.type),
+  }
+}
+```
+
+### SdkHttpResponse
+
+This is the response returned from an HTTP service. The emitters should take care to map it to what the method ultimately returns to end users.
+
+#### Interface
+
+```ts
+export interface SdkHttpResponse {
+  kind: "http";
+  headers: SdkServiceResponseHeader[];
+  apiVersions: string[];
+  type?: SdkType;
+  contentTypes: string[];
+  defaultContentType?: string;
+}
+```
+
+### SdkType
+
+These are the TCGC versions of each TypeSpec Type. They include information that all emitters want, i.e. their description directly on them.
+
+They all share this base:
+
+```ts
+interface SdkTypeBase {
+  // the raw TypeSpec type. Some of our types are currently
+  // created by tcgc. Those won't have an original type
+  __raw?: Type;
+  kind: string;
+  nullable: boolean;
```

Comment:
If a type is reference type, we could not set `nullable` to the type, instead need to set it to the property/parameter. We need to revisit the design of `nullable`.
```
model Test {
}
model A {
  prop: Test | null;
}
model B {
  prop: Test;
}

```

Reply:
Yeah, I agree with you. This was a discussion at some point in time, but I think nullable should actually be on the property / parameter instead. I'm going to update the design to this